pkg/model/api/v1/datasource/proxy.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2025 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datasource
+
+import (
+	"encoding/json"
+	"reflect"
+	"strings"
+)
+
+const (
+	ProxyKindField = "kind"
+	ProxySpec      = "spec"
+)
+
+// ValidateAndExtract finds a proxy in the pluginSpec
+// It then unmarshals the corresponding 'spec' field into the config interace{}.
+func ValidateAndExtract(pluginSpec interface{}) (interface{}, string, error) {
+	finder := &configFinder{}
+	finder.find(reflect.ValueOf(pluginSpec))
+
+	return finder.config, finder.foundKind, finder.err
+}
+
+type configFinder struct {
+	err       error
+	found     bool
+	foundKind string
+
+	config interface{}
+}
+
+func (c *configFinder) find(v reflect.Value) {
+	if c.err != nil || !v.IsValid() {
+		return
+	}
+
+	v = getNextElem(v)
+
+	switch v.Kind() {
+	case reflect.Struct:
+		c.findInStruct(v)
+	case reflect.Map:
+		c.findInMap(v)
+	}
+}
+
+// findInStruct searches for a kind/spec within a struct.
+func (c *configFinder) findInStruct(v reflect.Value) {
+	// first look at the field and find if we have the good kind
+	c.foundKind, c.found = getKindFromStruct(v)
+	if c.found {
+		// then get the spec field
+		spec := getConfigSpecInStruct(v)
+		// Then unmarshal the proxy to validate the content
+		c.unmarshalConfig(spec)
+	} else {
+		// Otherwise look deeper to find it
+		for i := 0; i < v.NumField(); i++ {
+			c.find(v.Field(i))
+			if c.found || c.err != nil {
+				return
+			}
+		}
+	}
+}
+
+// findInMap assumed that v is of type reflect.Map
+func (c *configFinder) findInMap(v reflect.Value) {
+	keyType := v.Type().Key()
+	if keyType.Kind() != reflect.String {
+		return
+	}
+
+	c.foundKind, c.found = getKindFromMap(v)
+	if c.found {
+		// then get the spec field
+		spec := getConfigSpecInMap(v)
+		// Then unmarshal the proxy to validate the content
+		c.unmarshalConfig(spec)
+	} else {
+		// Otherwise look deeper to find it
+		for _, key := range v.MapKeys() {
+			c.find(v.MapIndex(key))
+			if c.found || c.err != nil {
+				return
+			}
+		}
+	}
+}
+
+func (c *configFinder) unmarshalConfig(spec reflect.Value) {
+	if !spec.IsValid() {
+		return
+	}
+
+	spec = getNextElem(spec)
+	var data []byte
+	data, c.err = json.Marshal(spec.Interface())
```

Comment:
here we know the kind of config we need to un-marshal. So instead of marshalling and un-marshalling a generic struct, we can have the switch based on kind and unmarshal the correct config.

Doing that will also avoid to marshal and unmarshal twice the same data. A process that might have a cost when requesting 50/s the datasources from the server.